File: DataStructures/linkedList.cpp
#include <iostream>

using namespace std;

struct Node{
	int data;
	Node *next;
};

class LinkedList{
private:
	Node *first;
public:
	LinkedList(){
		first=NULL;
	}
	LinkedList(int A[], int n);
	~LinkedList();

	void Display();
	void Insert(int index, int x);
	int Delete(int index);
	int Length();
	void Search(int x);
};

//Constructor for LinkedList
LinkedList::LinkedList(int A[], int n){
	Node *last,*t;
	int i = 0;

	first = new Node;
	first->data = A[0];
	first->next = NULL;
	last = first;

	for(i = 1;i < n; i++){
		t = new Node;
		t->data = A[i];
		t->next = NULL;
		last->next = t;
		last = t;
	}
}

//Deconstructor for LinkedList
LinkedList::~LinkedList(){
	Node *p = first;
	while(first){
		first = first->next;
		delete p;
		p = first;
	}
}

//The time complexity is O(n) 
void LinkedList::Display(){
	Node *p = first;

	while(p){
		cout << p->data << " ";
		p = p->next;
	}

	cout << endl;
}

//The time complexity is O(n)
int LinkedList::Length(){
	Node *p = first;
	int len = 0;

	while(p){
		len++;
		p = p->next;
	}

	return len;
}

//The time complexity is O(x)
void LinkedList::Insert(int index, int x){
	Node *t,*p = first;

	if(index < 0 || index > Length())
		return;
	t = new Node;
	t->data = x;
	t->next = NULL;

	if(index == 0){
		t->next = first;
		first = t;
	}
	else{
		for(int i = 0; i < index; i++)
			p = p->next;
		t->next = p->next;
		p->next = t;
	}
}

//The time complexity is O(index)
int LinkedList::Delete(int index){
	Node *p, *q = NULL;
	int x = -1;

	if(index < 1 || index > Length())
		return -1;
	if(index == 1){
		p = first;
		first = first->next;
		x = p->data;
		delete p;
	}
	else{
		p = first;
		for(int i = 0; i < index - 1; i++){
			q = p;
			p = p->next;
		}
		q->next = p->next;
		x = p->data;
		delete p;
	}
	cout << x << endl;
	return x;
}

void LinkedList::Search(int x){
	//Start at the head and iterate through
	//Use a boolean value for found
	//Display message if the data can't be found
	Node *p = first;
	bool found = false;

	while(p && found == false){
		if(x == p->data){
			cout << "Found data " << x << endl;
			found = true;
		}
		else{
			p = p->next;
		}
	}

	if(found == false)
		cout << "The data " << x << " could not be found.\n";
}

int main(){
	
	int myArray[] = {1,2,3,4,5,6,7,8};
	LinkedList myLinkedList(myArray,8);

	myLinkedList.Display();
	myLinkedList.Insert(2,30);
	myLinkedList.Display();
	myLinkedList.Delete(4);
	myLinkedList.Display();

	myLinkedList.Search(20);
	myLinkedList.Search(7);

	return 0;
}

File: DataStructures/linkedList.py
'''
Now using modularized code with objects. Singly-Linked List in Python.
There are two classes - one for the Node and one for the head (LinkedList)
Functions include print (Methods of LinkedList), size, insert, search, and delete.
There are also getters and setters for the data and next node (methods of Node)
The time complexities are listed below. The space complexity is O(n)
'''

class LinkedList:
	def __init__(self):
		self.head = None

	#The time complexity of print is O(n)
	def print(self):
		current = self.head

		while current:
			print(current.data)
			current = current.nextNode

class Node:
	def __init__(self,data=None,nextNode=None):
		self.data = data
		self.nextNode = nextNode

	#The time complexity of the getters and setters is O(1)
	def get_data(self):
		return self.data

	def set_data(self,data):
		self.data = data

	def get_next(self):
		return self.nextNode

	def set_next(self,nextNode):
		self.nextNode = nextNode

#The time complexity of size is O(n)
def size(self):
	count = 0
	current = self.head

	while current:
		count = count + 1
		current = current.nextNode

	print(count)

#This inserts at the beginning
#The time complexity of print is O(1)
def insert(self,data):
	newNode = Node(data)
	newNode.set_next(self.head)
	self.head = newNode

#The time complexity of print is O(n) 
def search(self, data):	
	current = self.head
	found = False

	while current and found is False:
		if current.get_data() == data:
			found = True
		else:
			current = current.get_next()
	
	if current is None:
		raise ValueError("This node is not in the list.")

	return current

#The time complexity of print is O(n) 
def delete(self, data):

	current = self.head
	found = False
	previous = None

	while current and found is False:
		if current.get_data() == data:
			found = True
		else:
			previous = current
			current = current.get_next()

	if current is None:
		raise ValueError("This node is not in the list.")

	if previous is None:
		self.head = current.get_next()
	else:
		previous.set_next(current.get_next())

def main():
	
	myLinkedList = LinkedList()

	myLinkedList.head = Node(1)
	second = Node(2)
	third = Node(3) 

	myLinkedList.head.nextNode = second
	second.nextNode = third

	myLinkedList.print()
	insert(myLinkedList,4)
	size(myLinkedList)
	searchNode = search(myLinkedList,3)
	print(searchNode.data)
	delete(myLinkedList, 2)
	myLinkedList.print()


if __name__ == "__main__":
	main()